test/common/provider.c

```diff
@@ -0,0 +1,212 @@
+
+static enum umf_result_t traceAlloc(void *provider, size_t size,
+                                    size_t alignment, void **ptr) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("alloc");
+    return umfMemoryProviderAlloc(traceProvider->hUpstreamProvider, size,
+                                  alignment, ptr);
+}
+
+static enum umf_result_t traceFree(void *provider, void *ptr, size_t size) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("free");
+    return umfMemoryProviderFree(traceProvider->hUpstreamProvider, ptr, size);
+}
+
+static void traceGetLastError(void *provider, const char **ppMsg,
+                              int32_t *pError) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("get_last_native_error");
+    umfMemoryProviderGetLastNativeError(traceProvider->hUpstreamProvider, ppMsg,
+                                        pError);
+}
+
+static enum umf_result_t
+traceGetRecommendedPageSize(void *provider, size_t size, size_t *pageSize) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("get_recommended_page_size");
+    return umfMemoryProviderGetRecommendedPageSize(
+        traceProvider->hUpstreamProvider, size, pageSize);
+}
+
+static enum umf_result_t traceGetPageSize(void *provider, void *ptr,
+
+                                          size_t *pageSize) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("get_min_page_size");
+    return umfMemoryProviderGetMinPageSize(traceProvider->hUpstreamProvider,
+                                           ptr, pageSize);
+}
+
+static enum umf_result_t tracePurgeLazy(void *provider, void *ptr,
+                                        size_t size) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("purge_lazy");
+    return umfMemoryProviderPurgeLazy(traceProvider->hUpstreamProvider, ptr,
+                                      size);
+}
+
+static enum umf_result_t tracePurgeForce(void *provider, void *ptr,
+                                         size_t size) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("purge_force");
+    return umfMemoryProviderPurgeForce(traceProvider->hUpstreamProvider, ptr,
+                                       size);
+}
+
+static const char *traceName(void *provider) {
+    struct traceParams *traceProvider = (struct traceParams *)provider;
+
+    traceProvider->trace("name");
+    return umfMemoryProviderGetName(traceProvider->hUpstreamProvider);
+}
+
+umf_memory_provider_handle_t
+traceProviderCreate(umf_memory_provider_handle_t hUpstreamProvider,
+                    void (*trace)(const char *)) {
+    struct umf_memory_provider_ops_t ops = {
```

Comment:
... see my comment above - same for the tracing provider

Reply:
Done.